pkg/reconciler/ingress/ingress.go

```diff
@@ -0,0 +1,279 @@
+
+		rootIf, exists, err := c.indexer.Get(&networkingv1.Ingress{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:   ingress.Namespace,
+				Name:        rootIngressName,
+				ClusterName: ingress.GetClusterName(),
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		// TODO(jmprusi): A leaf without rootIngress?
+		if !exists {
+			return fmt.Errorf("Root Ingress not found: %s", rootIngressName)
+		}
+
+		rootIngress = rootIf.(*networkingv1.Ingress).DeepCopy()
+
+		// Clean the current status, and then recreate if from the other leafs.
+		rootIngress.Status.LoadBalancer.Ingress = []corev1.LoadBalancerIngress{}
+		for _, o := range others {
+			rootIngress.Status.LoadBalancer.Ingress = append(rootIngress.Status.LoadBalancer.Ingress, o.Status.LoadBalancer.Ingress...)
+		}
+
+		// If the envoy controlplane is enabled, we update the cache and generate and send to envoy a new snapshot.
+		if c.envoycontrolplane != nil {
+			err := c.envoycontrolplane.UpdateIngress(*rootIngress)
+			if err != nil {
+				return err
+			}
+
+			statusHost := generateStatusHost(c.domain, rootIngress)
+			// Now overwrite the Status of the rootIngress with our desired LB
+			rootIngress.Status.LoadBalancer.Ingress = []corev1.LoadBalancerIngress{{
+				Hostname: statusHost,
+			}}
+		}
+
+		// Update the rootIngress status with our desired LB.
+		if _, err := c.client.NetworkingV1().Ingresses(rootIngress.Namespace).UpdateStatus(ctx, rootIngress, metav1.UpdateOptions{}); err != nil {
+			if errors.IsConflict(err) {
+				key, err := cache.MetaNamespaceKeyFunc(ingress)
+				if err != nil {
+					return err
+				}
+				c.queue.AddRateLimited(key)
+				return nil
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+// desiredLeaves returns a list of leaves (ingresses) to be created based on an ingress.
+func (c *Controller) desiredLeaves(ctx context.Context, root *networkingv1.Ingress) ([]*networkingv1.Ingress, error) {
+	// This will parse the ingresses and extract all the destination services,
+	// then create a new ingress leaf for each of them.
+	services, err := c.getServices(ctx, root)
+	if err != nil {
+		return nil, err
+	}
+
+	var clusterDests []string
+	for _, service := range services {
+		if service.Labels[clusterLabel] != "" {
+			clusterDests = append(clusterDests, service.Labels[clusterLabel])
+		} else {
+			klog.Infof("Skipping service %q because it is not assigned to any cluster", service.Name)
+		}
+
+		// Trigger reconciliation of the root ingress when this service changes.
+		c.tracker.add(root, service)
+	}
+
+	if len(clusterDests) == 0 {
+		// No status conditions... let's just leave it blank for now.
+		root.Status.LoadBalancer.Ingress = []corev1.LoadBalancerIngress{{
+			IP:       "",
+			Hostname: "",
+		}}
+		return nil, nil
+	}
+
+	desiredLeaves := make([]*networkingv1.Ingress, 0, len(clusterDests))
+	for _, cl := range clusterDests {
+		vd := root.DeepCopy()
+		// TODO: munge cluster name
+		vd.Name = fmt.Sprintf("%s--%s", root.Name, cl)
```

Comment:
Let's resolve this now if you don't mind. We have the cluster name in the labels. We can generate a random name (e.g. uuid), or we can base it off of root.Name by e.g. setting GenerateName to `<base>-` where `<base>` is root.Name trimmed (if needed) so it doesn't exceed <max allowed name length - generated suffix length>, if that makes sense?